cookbook/tests/api/test_api_keyword.py

```diff
@@ -125,24 +190,189 @@ def test_add_duplicate(u1_s1, u1_s2, obj_1):
     response = json.loads(r.content)
     assert r.status_code == 201
     assert response['id'] != obj_1.id
+    assert json.loads(u1_s2.get(reverse(LIST_URL)).content)['count'] == 2
+
 
+def test_delete(u1_s1, u1_s2, obj_1, obj_1_1, obj_1_1_1):
+    with scopes_disabled():
+        assert Keyword.objects.count() == 3
 
-def test_delete(u1_s1, u1_s2, obj_1):
     r = u1_s2.delete(
         reverse(
             DETAIL_URL,
             args={obj_1.id}
         )
     )
     assert r.status_code == 404
+    with scopes_disabled():
+        assert Keyword.objects.count() == 3
 
     r = u1_s1.delete(
         reverse(
             DETAIL_URL,
-            args={obj_1.id}
+            args={obj_1_1.id}
         )
     )
 
     assert r.status_code == 204
     with scopes_disabled():
-        assert Keyword.objects.count() == 0
+        assert Keyword.objects.count() == 1
+        assert Keyword.find_problems() == ([], [], [], [], [])
+
+
+def test_move(u1_s1, obj_1, obj_1_1, obj_1_1_1, obj_2, obj_3, space_1):
+    url = reverse(MOVE_URL, args=[obj_1_1.id, obj_2.id])
+    with scopes_disabled():
+        assert obj_1.get_num_children() == 1
+        assert obj_1.get_descendant_count() == 2
+        assert Keyword.get_root_nodes().filter(space=space_1).count() == 2
+
+    # move child to new parent, only HTTP put method should work
+    r = u1_s1.get(url)
+    assert r.status_code == 405
+    r = u1_s1.post(url)
+    assert r.status_code == 405
+    r = u1_s1.delete(url)
+    assert r.status_code == 405
+    r = u1_s1.put(url)
+    assert r.status_code == 200
+    with scopes_disabled():
+        # django-treebeard bypasses django ORM so object needs retrieved again
+        obj_1 = Keyword.objects.get(pk=obj_1.id)
+        obj_2 = Keyword.objects.get(pk=obj_2.id)
+        assert obj_1.get_num_children() == 0
+        assert obj_1.get_descendant_count() == 0
+        assert obj_2.get_num_children() == 1
+        assert obj_2.get_descendant_count() == 2
+
+    # move child to root
+    r = u1_s1.put(reverse(MOVE_URL, args=[obj_1_1.id, 0]))
+    assert r.status_code == 200
+    with scopes_disabled():
+        assert Keyword.get_root_nodes().filter(space=space_1).count() == 3
+
+    # attempt to move to non-existent parent
+    r = u1_s1.put(
+        reverse(MOVE_URL, args=[obj_1.id, 9999])
+    )
+    assert r.status_code == 400
+
+    # attempt to move to wrong space
+    r = u1_s1.put(
+        reverse(MOVE_URL, args=[obj_1_1.id, obj_3.id])
+    )
+    assert r.status_code == 400
```

Comment:
should this call return 404 ? most space related api calls return 404 when attempting to use an object that does not belong to ones space in order to prevent leaking information about the existente or non existente of certain objects.

Reply:
404 is fine - I didn't realize you had a convention in place.

Reply:
technically using incrementing IDs is already a bad practice but i prefer them over guids but that means that one should not be able to verify the existence of objects by incrementing ids